### PR TITLE
Lower convolution to KRNL dialect.

### PR DESCRIPTION
**Pull request by doru1004**
Path that lowers convolution to KRNL dialect.

### CODE REVIEW

**Comment by doru1004** [comment]
@tungld I have made some refactoring here based on the repetitive patterns I spotted when implementing convolution. I also changes some of the code in the templates since it was too similar. Please let me know if you agree with the changes. I know you have a patch which performs some code simplification in this file. Maybe you can simplify this code further once Conv lands?

**Comment by tungld** [comment]
The changes look good to me. Just a small comment about insertion for optimization, since we have no optimization, we can put those parts inside `defineLoops`.

File: src/pass/lower_frontend_to_krnl.cpp [hunk 1]
```diff
@@ -1638,6 +1574,255 @@ struct ONNXIdentityOpLowering : public ConversionPattern {
...
+      for (int i = 2; i < resultShape.size(); ++i)
+        addDimensionToPack(rewriter, loc, spatialPack, alloc, i);
+
+      // 2.4 Emit loop nest over output spatial dimensions.
+      //   for rX = 0 .. RX
+      auto spatialIterateOp =
+          rewriter.create<KrnlIterateOp>(loc, spatialPack);
+      Block &spatialIterationBlock = spatialIterateOp.bodyRegion().front();
+      // 2.5 Emit optimizations for outer loops:
+      rewriter.setInsertionPointToEnd(optSpatialLoopBlock);
```
Comment (tungld) [inline src/pass/lower_frontend_to_krnl.cpp]: How about putting the following two lines inside function `defineLoops`?

File: src/pass/lower_frontend_to_krnl.cpp [hunk 2]
```diff
@@ -1638,6 +1574,255 @@ struct ONNXIdentityOpLowering : public ConversionPattern {
...
+        //   for Kx = 0 .. KX
+        for (int i = 2; i < kernelShape.size(); ++i)
+          addDimensionToPack(rewriter, loc, innerPack, operands[1], i);
+
+        // 3.4 Emit inner loop nest.
+        auto innerIterateOp =
+            rewriter.create<KrnlIterateOp>(loc, innerPack);
+        Block &innerIterationBlock = innerIterateOp.bodyRegion().front();
+        // 3.5 Emit optimizations for outer loops:
+        rewriter.setInsertionPointToEnd(optInnerLoopBlock);
```
Comment (tungld) [inline src/pass/lower_frontend_to_krnl.cpp]: How about putting the following two lines inside function `defineLoops`?

**Comment by doru1004** [comment]
> The changes look good to me. Just a small comment about insertion for optimization, since we have no optimization, we can put those parts inside `defineLoops`.

I thought about that but it doesn't conceptually belong inside the function since we might want to actually add optimizations which will be on a per operation basis. When that happens we will no longer return the un-optimized loops like we do today so the return statement will have to be changed by every op.

File: src/pass/lower_frontend_to_krnl.cpp [hunk 3]
```diff
@@ -1638,6 +1574,255 @@ struct ONNXIdentityOpLowering : public ConversionPattern {
...
+
+    if (hasAllConstantDimensions(memRefType))
+      alloc = insertAllocAndDealloc(memRefType, loc, rewriter, insertDealloc);
+    else
+      alloc = insertAllocAndDealloc(memRefType, loc, rewriter, insertDealloc,
+                                    {operands[0]});
+
+    auto resultShape = memRefType.getShape();
+    auto inputShape = operands[0].getType().cast<MemRefType>().getShape();
+    auto kernelShape = operands[1].getType().cast<MemRefType>().getShape();
```
[the remark this answers is not in the document]
Comment (doru1004) [reply src/pass/lower_frontend_to_krnl.cpp]: It seems that the operands can only be used from the list of operands that is passed as input to the matchAndRewrite() function.

File: src/pass/lower_frontend_to_krnl.cpp [hunk 4]
```diff
@@ -1638,6 +1574,255 @@ struct ONNXIdentityOpLowering : public ConversionPattern {
...
+    // of r- and k-index loop i.e. r1 r2 and k1 k2 loops.
+
+    // Set up outermost loops: n g m r1 r2 ... rdim
+    // Skip g if group is 1.
+
+    // Before we start the iteration we need to compute the number of
+    // unsplit kernels and fetch the number of groups from the attribute
+    // list. Group is always a compilation constant.
+    int64_t group =
+        llvm::dyn_cast<ONNXConvNoBiasOp>(op).group().getSExtValue();
```
[the remark this answers is not in the document]
Comment (doru1004) [reply src/pass/lower_frontend_to_krnl.cpp]: Done

**Pull request merged**